README.md
restartd - A process-restarting daemon
======================================

This is a new upstream for the restartd(8) daemon, which was originally
debian-only. It will host the new developments for this project.

From the man page:

    Restartd is a daemon for checking running and not running processes. It
    reads the /proc directory every n seconds and does a POSIX regexp
    on the process names. You can execute a script or a program if the
    process is or is not running.


Links
-----
* [Debian QA page for the restartd package](http://packages.qa.debian.org/r/restartd.html)
* [Debian packaging repository for the restartd package](https://github.com/ajraymond/restartd-debian)
* [Ubuntu bug tracker for the restartd package](https://launchpad.net/ubuntu/+source/restartd/+bugs)


CHANGELOG
restartd - A process-restarting daemon - Changelog
==================================================

What's new in v0.2.4:
  * Rename main.c to restartd.c
  * Update makefile to support debuild harding options

What's new in v0.2.3
  * Project now has a standalone upstream
  * Rename LICENCE to LICENSE


main.c
// restartd - Process checker and/or restarter daemon
// Copyright (C) 2000-2002 Tibor Koleszar <oldw@debian.org>
// Copyright (C) 2006 Aurélien GÉRÔME <ag@roxor.cx>
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
//
// Tibor Koleszar <oldw@debian.org>
// H-7400 Kaposvar, Somssich Pal str. 4.

#include <sys/types.h>
#include <dirent.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <regex.h>
#include <time.h>
#include <syslog.h>
#include <signal.h>
#include <stdarg.h>

#include "config.h"

// SIGTERM & SIGHUP handler
void
got_signal (int sig)
{

  if (sig == SIGTERM)
    {
      syslog (LOG_INFO, "Got SIGTERM. Exiting.");
      unlink ("/var/run/restartd.pid");
      exit (0);
    }
  if (sig == SIGHUP)
    {
      syslog (LOG_INFO, "Got SIGHUP. Reloading config file.");
      read_config ();
      syslog (LOG_INFO,
	      "Config file has been read. Found %d process to check.",
	      config_process_number);
    }

}


int
main (int argc, char *argv[])
{

  DIR *procdir_id;
  struct dirent *procdir_dirent;
  int proc_cmdline;
  char *proc_cmdline_str;
  ssize_t proc_cmdline_str_length;
  char *proc_cmdline_name;
  int i, f;
  FILE *out_proc;
  regex_t *regc;
  size_t r_nmatch = 0;
  regmatch_t r_match[256];
  pid_t child;
  time_t now;

  debug = 0;
  check_interval = 20;
  foreground = 0;

  // Options

  config_file = strdup (DEFAULT_CONFIG);

  for (i = 0; i < argc; i++)
    {

      if (!strcmp (argv[i], "-c") || !strcmp (argv[i], "--config"))
	config_file = strdup (argv[i + 1]);

      if (!strcmp (argv[i], "-v") || !strcmp (argv[i], "--version"))
	{
	  printf
	    ("restard %s - Copyright 2000-2002 Tibor Koleszar <oldw@debian.org>\n"
	     "                Copyright 2006 Aurélien GÉRÔME <ag@roxor.cx>\n",
	     VERSION);
	  exit (0);
	}

      if (!strcmp (argv[i], "-d") || !strcmp (argv[i], "--debug"))
	debug = 1;

      if (!strcmp (argv[i], "-f") || !strcmp (argv[i], "--foreground"))
	foreground = 1;

      if (!strcmp (argv[i], "-i") || !strcmp (argv[i], "--interval"))
	{
	  if (atoi (argv[i + 1]) > 0)
	    {
	      check_interval = atoi (argv[i + 1]);
	    }
	  else
	    {
	      fprintf (stderr,
		       "Error at parameter -i: %s is not an interval number\n.",
		       argv[i + 1]);
	      exit (0);
	    }
	}

      if (!strcmp (argv[i], "-h") || !strcmp (argv[i], "--help"))
	{
	  printf ("restartd %s - process restarter and checker daemon\n"
		  "Copyright 2000-2002 Tibor Koleszar <oldw@debian.org>\n"
		  "Copyright 2006 Aurélien GÉRÔME <ag@roxor.cx>\n"
		  "restartd is GPL'ed software.\n"
		  "Parameters:\n"
		  "  -c <config_file> : config file (default is /etc/restartd.conf)\n"
		  "  -d               : set debug on (poor)\n"
		  "  -f               : run in foreground\n"
		  "  -i <interval_sec>: the check interval in second\n"
		  "  -h               : help\n\n", VERSION);
	}

    }

  config_process = malloc (sizeof (struct config_process_type) * 128);

  read_config ();

  syslog (LOG_INFO, "Config file has been read. Found %d process to check.",
	  config_process_number);

  procdir_dirent = malloc (sizeof (struct dirent));
  proc_cmdline_str = (char *) malloc (1024);
  proc_cmdline_name = (char *) malloc (1024);
  regc = malloc (1024);

  // Catch signals
  signal (SIGTERM, got_signal);
  signal (SIGHUP, got_signal);

  // Fork
  if (foreground == 1 || !(child = fork ()))
    {

      // generic daemon stuff from http://www.aarnet.edu.au/~gdt/articles/2003-09-nowhere-daemon/daemon.html
      // added by cwarden@postica.com - 041216
      if (!foreground)
	{
	  if (setsid () == (pid_t) (-1))
	    {
	      fprintf (stderr, "setsid() failed\n");
	      syslog (LOG_ERR, "setsid() failed");
	      return -1;
	    }
	}
      if ((chdir ("/")) < 0)
	{
	  syslog (LOG_ERR, "Cannot chdir to /");
	  return -1;
	}
      f = open ("/dev/null", O_RDONLY);
      if (f == -1)
	{
	  syslog (LOG_ERR, "Failed to open /dev/null");
	  return -1;
	}
      else
	{
	  if (dup2 (f, fileno (stdin)) == -1)
	    {
	      syslog (LOG_ERR, "Failed to set stdin to /dev/null");
	      return -1;
	    }
	  (void) close (f);
	}

      f = open ("/dev/null", O_WRONLY);
      if (f == -1)
	{
	  syslog (LOG_ERR, "Failed to open /dev/null");
	  return -1;
	}
      else
	{
	  if (dup2 (f, fileno (stdout)) == -1)
	    {
	      syslog (LOG_ERR, "Failed to set stdout to /dev/null");
	      return -1;
	    }
	  if (dup2 (f, fileno (stderr)) == -1)
	    {
	      syslog (LOG_ERR, "Failed to set stderr to /dev/null");
	      return -1;
	    }
	  (void) close (f);
	}

      out_proc = fopen ("/var/run/restartd.pid", "wt");
      fprintf (out_proc, "%d", getpid ());
      fclose (out_proc);

      while (1)
	{

	  if ((procdir_id = opendir ("/proc")) == NULL)
	    {
	      syslog (LOG_ERR, "Cannot open /proc dir");
	      return -1;
	    }

	  for (i = 0; i < config_process_number; i++)
	    config_process[i].processes[0] = 0;

	  while ((procdir_dirent = readdir (procdir_id)) != NULL)

	    if (atoi (procdir_dirent->d_name) > 0)
	      {

		sprintf (proc_cmdline_name, "/proc/%s/cmdline",
			 procdir_dirent->d_name);
		proc_cmdline = open (proc_cmdline_name, O_RDONLY);

		proc_cmdline_str_length =
		  read (proc_cmdline, proc_cmdline_str, 1024);

		for (i = 0; i < proc_cmdline_str_length; i++)
		  if (!proc_cmdline_str[i])
		    proc_cmdline_str[i] = 32;

		proc_cmdline_str[proc_cmdline_str_length] = 0;

		// Check process
		for (i = 0; i < config_process_number; i++)
		  {
		    regcomp (regc, config_process[i].regexp, REG_EXTENDED
			     || REG_ICASE);
		    if (!regexec
			(regc, proc_cmdline_str, r_nmatch, r_match, 0))
		      sprintf (config_process[i].processes, "%s %s",
			       config_process[i].processes,
			       procdir_dirent->d_name);
		    regfree (regc);
		  }

		close (proc_cmdline);

	      }

	  closedir (procdir_id);

	  now = time (NULL);

	  out_proc = fopen ("/var/run/restartd", "wt");

	  fprintf (out_proc, "%s\n", ctime (&now));
	  for (i = 0; i < config_process_number; i++)
	    {

	      if (strlen (config_process[i].processes) > 0)
		{
		  if (strlen (config_process[i].running) > 0)
		    {
		      strcpy (config_process[i].status, "running");
		      syslog (LOG_INFO, "%s is running, executing '%s'",
			      config_process[i].name,
			      config_process[i].running);
		      system (config_process[i].running);
		    }
		  else
		    {
		      strcpy (config_process[i].status, "running");
		    }
		}

	      if (strlen (config_process[i].processes) == 0
		  && strlen (config_process[i].not_running) > 0)
		{
		  strcpy (config_process[i].status, "restarting");
		  syslog (LOG_ERR, "%s is NOT running, executing '%s'",
			  config_process[i].name,
			  config_process[i].not_running);
		  system (config_process[i].not_running);
		}

	      if (strlen (config_process[i].processes) == 0
		  && strlen (config_process[i].not_running) == 0)
		strcpy (config_process[i].status, "not running");

	      fprintf (out_proc, "%-12s %-12s      %s\n",
		       config_process[i].name, config_process[i].status,
		       config_process[i].processes);
	    }

	  fclose (out_proc);

	  sleep (check_interval);

	}
    }

  free (procdir_dirent);
  free (proc_cmdline_str);
  free (proc_cmdline_name);
  free (config_process);
  free (regc);
  return 0;

}


config.c
/* restartd - Process checker and/or restarter daemon
 * Copyright (C) 2000-2002 Tibor Koleszar <oldw@debian.org>
 * Copyright (C) 2006 Aurélien GÉRÔME <ag@roxor.cx>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * Tibor Koleszar <oldw@debian.org>
 * H-7400 Kaposvar, Somssich Pal str. 4.
 */

#include <syslog.h>
#include <sys/types.h>
#include <dirent.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <fcntl.h>

#include "config.h"

#define MAX_LINE_LENGTH 1024

int read_config(void)
{
    FILE *config_fd;
    char *line1, *line2;
    int i, j, sep, lines;

    if (debug) {
        fprintf(stderr, "Config file: %s\n", config_file);
    }

    /* This is a reload */
    if (config_process_number) {
        for(i=0; i<config_process_number; i++) {
            config_process[i].name[0] = 0;
            config_process[i].regexp[0] = 0;
            config_process[i].running[0] = 0;
            config_process[i].not_running[0] = 0;
        }
    }

    config_process_number = 0;

    line1 = (char *) malloc(MAX_LINE_LENGTH);
    line2 = (char *) malloc(MAX_LINE_LENGTH);

    if ((config_fd = fopen(config_file, "rt")) == NULL) {
        fprintf(stderr, "Error at opening config file: %s\n", config_file);
        syslog(LOG_ERR, "Error at opening config file: %s", config_file);
        _exit(0);
    }

    while(fgets(line1, MAX_LINE_LENGTH, config_fd) != NULL) {
        j = 0;

        for(i=0; i<strlen(line1); i++) {
            if (line1[i] == '\n' || line1[i] == '\r' || line1[i] == '\t'
                || line1[i] == ' ') {
                if (line2[j - 1] != ' ' && j > 0) {
                    line2[j] = ' ';
                    j++;
                }
            } else {
                line2[j] = line1[i];
                j++;
            }
        }

        line2[j] = 0;

        if (line2[0] != '#' && strlen(line2) > 0) {
            j = 0;
            i = 0;
            lines = 0;
            sep = 0;

            while(j < strlen(line2)) {
                if (line2[j] == ' ' && sep != 1) {
                    line1[i] = 0;

                    if (debug) {
                        fprintf(stderr, "%u: %s\n", lines, line1);
                    }

                    i = 0;
                    sep = 0;

                    if (lines == 0) {
                        snprintf(config_process[config_process_number].name,
                                 64, "%s", line1);
                    }
                    if (lines == 1) {
                        snprintf(config_process[config_process_number].regexp,
                                 128, "%s", line1);
                    }
                    if (lines == 2) {
                        snprintf(config_process[config_process_number].not_running,
                                 256, "%s", line1);
                    }
                    if (lines == 3) {
                        snprintf(config_process[config_process_number].running,
                                 256, "%s", line1);
                    }

                    lines++;
                }

                if (line2[j] == '"') {
                    sep++;
                }

                if ((line2[j] != ' ' || sep == 1) && line2[j] != '"') {
                    line1[i] = line2[j];
                    i++;
                }

                j++;
            }

            config_process_number++;
        }

    }

    fclose(config_fd);

    free(line1);
    free(line2);

    if (!config_process_number) {
        fprintf(stderr, "No processes defined in config file. Exiting.\n");
        syslog(LOG_INFO, "No processes defined in config file. Exiting.");
        exit(0);
    }

    return 0;
}

void dump_config(void) {
    int i;

    for(i=0; i<config_process_number; i++) {
        printf("ID=%d\n  name=%s\n  regexp=%s\n  running=%s\n  not_running=%s\n", i,
            config_process[i].name,
            config_process[i].regexp,
            config_process[i].running,
            config_process[i].not_running);
    }

}


config.h
/* restartd - Process checker and/or restarter daemon
 * Copyright (C) 2000-2002 Tibor Koleszar <oldw@debian.org>
 * Copyright (C) 2006 Aurélien GÉRÔME <ag@roxor.cx>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * Tibor Koleszar <oldw@debian.org>
 * H-7400 Kaposvar, Somssich Pal str. 4.
 */

#ifndef RESTARTD_CONFIG_H
#define RESTARTD_CONFIG_H

#define DEFAULT_CONFIG "/etc/restartd.conf"

int debug;
int config_process_number;
int check_interval;
int foreground;
struct config_process_type *config_process;
char *config_file;

typedef struct config_process_type {
  char name[64];
  char regexp[128];
  char not_running[256];
  char running[256];
  char processes[256];
  char status[32];
} config_process_type;

int read_config(/* char *config_file */);
void dump_config(void);

#endif /* RESTARTD_CONFIG_H */


restartd.c
/* restartd - Process checker and/or restarter daemon
 * Copyright (C) 2000-2002 Tibor Koleszar <oldw@debian.org>
 * Copyright (C) 2006 Aurélien GÉRÔME <ag@roxor.cx>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * Tibor Koleszar <oldw@debian.org>
 * H-7400 Kaposvar, Somssich Pal str. 4.
 */
#include <sys/types.h>
#include <dirent.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <regex.h>
#include <time.h>
#include <syslog.h>
#include <signal.h>
#include <stdarg.h>

#include "config.h"

/* SIGTERM & SIGHUP handler */
void got_signal(int sig)
{
    if (sig == SIGTERM) {
        syslog(LOG_INFO, "Got SIGTERM. Exiting.");
        unlink("/var/run/restartd.pid");
        exit(0);
    }
    if (sig == SIGHUP) {
        syslog(LOG_INFO, "Got SIGHUP. Reloading config file.");
        read_config();
        syslog(LOG_INFO,
               "Config file has been read. Found %d process to check.",
               config_process_number);
    }
}

int main(int argc, char *argv[])
{
    DIR *procdir_id;
    struct dirent *procdir_dirent;
    int proc_cmdline;
    char *proc_cmdline_str;
    ssize_t proc_cmdline_str_length;
    char *proc_cmdline_name;
    int i, f;
    FILE *out_proc;
    regex_t *regc;
    size_t r_nmatch = 0;
    regmatch_t r_match[256];
    pid_t child;
    time_t now;
    int list_only;

    debug = 0;
    check_interval = 20;
    foreground = 0;

    /* Options */
    config_file = strdup(DEFAULT_CONFIG);
    list_only = 0;

    for(i = 0; i < argc; i++) {
        if (!strcmp(argv[i], "-c") || !strcmp(argv[i], "--config")) {
            config_file = strdup(argv[i + 1]);
        }
        if (!strcmp(argv[i], "-v") || !strcmp(argv[i], "--version")) {
            printf("restard %s - Copyright 2000-2002 Tibor Koleszar <oldw@debian.org>\n"
                   "                Copyright 2006 Aurélien GÉRÔME <ag@roxor.cx>\n",
                   VERSION);
            exit(0);
        }
        if (!strcmp(argv[i], "-d") || !strcmp(argv[i], "--debug")) {
            debug = 1;
        }
        if (!strcmp(argv[i], "-f") || !strcmp(argv[i], "--foreground")) {
            foreground = 1;
        }
        if (!strcmp(argv[i], "-i") || !strcmp(argv[i], "--interval")) {
            if (atoi(argv[i + 1]) > 0) {
                check_interval = atoi(argv[i + 1]);
            } else {
                fprintf(stderr,
                        "Error at parameter -i: %s is not an interval number\n.",
                         argv[i + 1]);
                exit (0);
            }
        }
        if (!strcmp(argv[i], "-l") || !strcmp(argv[i], "--list")) {
            list_only = 1;
        }
        if (!strcmp(argv[i], "-h") || !strcmp(argv[i], "--help")) {
            printf("restartd %s - process restarter and checker daemon\n"
                    "Copyright 2000-2002 Tibor Koleszar <oldw@debian.org>\n"
                    "Copyright 2006 Aurélien GÉRÔME <ag@roxor.cx>\n"
                    "restartd is GPL'ed software.\n"
                    "Parameters:\n"
                    "  -c <config_file> : config file (default is /etc/restartd.conf)\n"
                    "  -d               : set debug on (poor)\n"
                    "  -f               : run in foreground\n"
                    "  -i <interval_sec>: the check interval in second\n"
                    "  -l               : list configuration options\n"
                    "  -h               : help\n\n", VERSION);
        }
    }

    config_process = malloc(sizeof(struct config_process_type) * 128);
  
    read_config();
    if (list_only) {
        dump_config();
        exit(0);
    }
  
    syslog(LOG_INFO, "Config file has been read. Found %d process to check.",
           config_process_number);
  
    procdir_dirent = malloc(sizeof(struct dirent));
    proc_cmdline_str = (char *) malloc(1024);
    proc_cmdline_name = (char *) malloc(1024);
    regc = malloc(1024);
  
    /* Catch signals */
    signal(SIGTERM, got_signal);
    signal(SIGHUP, got_signal);

    /* Fork */
    if (foreground == 1 || !(child = fork())) {
        /* generic daemon stuff from http://www.aarnet.edu.au/~gdt/articles/2003-09-nowhere-daemon/daemon.html
           added by cwarden@postica.com - 041216 */
        if (!foreground) {
            if (setsid() == (pid_t) (-1)) {
                fprintf(stderr, "setsid() failed\n");
                syslog(LOG_ERR, "setsid() failed");
                return -1;
            }
        }

        if ((chdir("/")) < 0) {
            syslog(LOG_ERR, "Cannot chdir to /");
            return -1;
        }

        f = open("/dev/null", O_RDONLY);
        if (f == -1) {
            syslog(LOG_ERR, "Failed to open /dev/null");
            return -1;
        } else {
            if (dup2(f, fileno(stdin)) == -1) {
                syslog(LOG_ERR, "Failed to set stdin to /dev/null");
                return -1;
            }
            (void) close(f);
        }

        f = open("/dev/null", O_WRONLY);
        if (f == -1) {
            syslog(LOG_ERR, "Failed to open /dev/null");
            return -1;
        } else {
            if (dup2(f, fileno(stdout)) == -1) {
                syslog(LOG_ERR, "Failed to set stdout to /dev/null");
                return -1;
            }
            if (dup2(f, fileno(stderr)) == -1) {
                syslog(LOG_ERR, "Failed to set stderr to /dev/null");
                return -1;
            }
            (void) close(f);
        }

        out_proc = fopen("/var/run/restartd.pid", "wt");
        fprintf(out_proc, "%d", getpid());
        fclose(out_proc);

        while(1) {
            if ((procdir_id = opendir("/proc")) == NULL) {
                syslog(LOG_ERR, "Cannot open /proc dir");
                return -1;
            }

            for(i = 0; i < config_process_number; i++) {
                config_process[i].processes[0] = 0;
            }

            while((procdir_dirent = readdir(procdir_id)) != NULL) {
                if (atoi(procdir_dirent->d_name) > 0) {
                    sprintf(proc_cmdline_name, "/proc/%s/cmdline",
                            procdir_dirent->d_name);
                    proc_cmdline = open(proc_cmdline_name, O_RDONLY);
         
                    proc_cmdline_str_length =
                        read(proc_cmdline, proc_cmdline_str, 1024);
         
                    for(i = 0; i < proc_cmdline_str_length; i++) {
                        if (!proc_cmdline_str[i]) {
                            proc_cmdline_str[i] = 32;
                        }
                    }
         
                    proc_cmdline_str[proc_cmdline_str_length] = 0;
         
                    /* Check process */
                    for(i=0; i<config_process_number; i++) {
                        regcomp(regc, config_process[i].regexp, REG_EXTENDED || REG_ICASE);
                        if (!regexec(regc, proc_cmdline_str, r_nmatch, r_match, 0)) {
                            sprintf(config_process[i].processes, "%s %s",
                                    config_process[i].processes,
                                    procdir_dirent->d_name);
                        }
                        regfree(regc);
                    }
         
                    close(proc_cmdline);
                }
            }

           closedir(procdir_id);

           now = time(NULL);

           out_proc = fopen("/var/run/restartd", "wt");

           fprintf(out_proc, "%s\n", ctime(&now));

           for(i=0; i<config_process_number; i++) {
               if (strlen(config_process[i].processes) > 0) {
                   if (strlen(config_process[i].running) > 0) {
                       strcpy(config_process[i].status, "running");
                       syslog(LOG_INFO, "%s is running, executing '%s'",
                              config_process[i].name,
                              config_process[i].running);
                       system(config_process[i].running);
                    } else {
                        strcpy(config_process[i].status, "running");
                    }
                }

                if (strlen(config_process[i].processes) == 0
                    && strlen(config_process[i].not_running) > 0) {
                    strcpy(config_process[i].status, "restarting");
                    syslog(LOG_ERR, "%s is NOT running, executing '%s'",
                           config_process[i].name,
                           config_process[i].not_running);
                    system(config_process[i].not_running);
                }

                if (strlen(config_process[i].processes) == 0
                    && strlen(config_process[i].not_running) == 0) {
                    strcpy(config_process[i].status, "not running");
                }

                fprintf(out_proc, "%-12s %-12s      %s\n",
                        config_process[i].name, config_process[i].status,
                        config_process[i].processes);
            }

            fclose(out_proc);

            sleep(check_interval);
        }
    }

    free(procdir_dirent);
    free(proc_cmdline_str);
    free(proc_cmdline_name);
    free(config_process);
    free(regc);

    return 0;
}

